python/tank_vendor/shotgun_authentication/interactive_authentication.py

```diff
@@ -330,50 +349,59 @@ def _process_ui():
+
+class AuthState:
+    """
+    Authentication can be in three states. Waiting for authentication to complete,
+    authentication was cancelled (user hit Cancel or CTRL-D) and user provided
+    credentials successfully.
+    """
+    waiting = 1
+    cancelled = 2
+    success = 3
 # When a thread cancels session renewal, this flag is set so other threads know
 # to raise an exception as well.
-_is_authentication_cancelled = False
+_auth_state = AuthState.waiting
 
 
-def _renew_session_internal(user, session_token, credentials_handler):
+def _renew_session_internal(user, credentials_handler):
     """
-    Common login logic, regardless of how we are actually logging in. It will first try to reuse
-    any existing session and if that fails then it will ask for credentials and upon success
-    the credentials will be cached.
-    :raises AuthenticationError: Raised if the authentication is cancelled.
-    :raises AuthenticationDisabled: Raised if authentication was cancelled before.
+    Prompts the user for the password. This method should never be called directly
+    and _renew_session should be called instead.
+    :raises AuthenticationCancelled: Raised if the authentication is cancelled.
     """
     logger.debug("About to take the authentication lock.")
     global _renew_session_internal_lock
     with _renew_session_internal_lock:
-        global _is_authentication_cancelled
-        if _is_authentication_cancelled:
-            raise AuthenticationCancelled()
 
-        # If somebody refreshed the session token on the user object since we tried with
-        # the session token.
-        if user.get_session_token() != session_token:
-            return None
         logger.debug("Took the authentication lock.")
 
+        # When authentication is cancelled, every thread who enter the authentication
+        # critical section should throw as well.
+        global _auth_state
+        if _auth_state == AuthState.cancelled:
+            raise AuthenticationCancelled()
+        # If authentication was successful, simply return.
+        elif _auth_state == AuthState.success:
+            return
+
+        # We're the first thread, so authenticate.
         try:
             logger.debug("Not authenticated, requesting user input.")
-            # Do the actually credentials prompting and authenticating.
             hostname, login, session_token = credentials_handler.authenticate(
                 user.get_host(),
                 user.get_login(),
                 user.get_http_proxy()
             )
+            _auth_state = AuthState.success
+            logger.debug("Login successful!")
+            user.set_session_token(session_token)
         except AuthenticationCancelled:
-            _is_authentication_cancelled = True
+            _auth_state = AuthState.cancelled
             logger.debug("Authentication cancelled")
-            user.clear_saved_user(user.get_host())
+            user.uncache_session_token()
```

Comment:
Sorry, one more thought - `uncache` sounds a little unusual. If we have to keep this method, I suggest renaming to `clear_cached_session_token`